### ISSUE TITLE
Check lints related to trait implementations for false-positives if the trait impl is `#[doc(hidden)]`

### ISSUE THREAD

**Issue by obi1kenobi** [obi1kenobi/cargo-semver-checks#762]
For example:
```rust
pub struct SecretlyIter;

#[doc(hidden)]
impl Iterator for SecretlyIter {
    type Item = i64;

    fn next(&mut self) -> Option<Self::Item> {
        None
    }
}
```

The fact that `SecretlyIter` is `impl Iterator` is not public API. But there is still a hazard, since users may rely on `SecretlyIter` being `impl Iterator` indirectly, e.g. by using it in a `for` loop. Discussion on that here: https://hachyderm.io/@kornel@mastodon.social/112292583897196920

This situation will require some nuance. Tentative plan:
- [ ] Add test cases for various edge cases, to pin down the current behavior of our lints.
- [ ] Make the "trait no longer implemented" lint check that the original impl was public API.
- [ ] Add a new lint for "trait impl stopped being public API".
- [ ] After #58, add a new warning-level lint that checks for "trait no longer implemented" _ignoring_ the public API property, narrowed specifically for special traits in the Rust prelude like `Iterator`.
- [ ] Eliminate any other false-positives, and consider adding more lints if needed, analogously to the above.
